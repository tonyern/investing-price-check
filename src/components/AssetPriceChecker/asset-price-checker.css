table,
th,
td {
  border: 1px solid black;
}

table {
  width: 100%;
}

th,
td {
  width: 50%;
}

.asset-price-box {
  text-align: center;
}

.asset-price-box .title {
  font-size: 48px;
  font-weight: 1000;
}

.asset-price-box .asset {
  font-size: 20px;
  margin: 10px 0px 10px 0px;
}

.price-btn {
  padding: 0px 130px 0px 130px;
}

.price-btn .price-btn-title {
  font-size: 32px;
}
