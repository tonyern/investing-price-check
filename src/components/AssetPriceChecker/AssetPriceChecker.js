import React, { useState } from "react";
import axios from "axios";
import "./asset-price-checker.css";

const AssetPriceChecker = () => {
  const metalsAPI = {
    key: "bd8x58t67kikf2jueqzeex2g0hpysu9rbzlgm9vbu3of34yhla013s04dz3r",
    base: "https://www.metals-api.com/api/latest",
    currency: "USD",
  };

  const [bitcoinPrice, setBitcoinPrice] = useState("---");
  const [bitcoinCashPrice, setBitcoinCashPrice] = useState("---");
  const [goldPrice, setGoldPrice] = useState("---");
  const [silverPrice, setSilverPrice] = useState("---");

  const dateBuilder = (d) => {
    let months = [
      "January",
      "February",
      "March",
      "April",
      "May",
      "June",
      "July",
      "August",
      "September",
      "October",
      "November",
      "December",
    ];
    let days = [
      "Sunday",
      "Monday",
      "Tuesday",
      "Wednesday",
      "Thursday",
      "Friday",
      "Saturday",
    ];

    let day = days[d.getDay()];
    let date = d.getDate();
    let month = months[d.getMonth()];
    let year = d.getFullYear();

    return `${day}, ${date} ${month} ${year}`;
  };

  const fetchAPI = () => {
    axios
      .get(
        `${metalsAPI.base}?access_key=${metalsAPI.key}&base=${metalsAPI.currency}`
      )
      .then((response) => {
        console.log(response.data);
        setBitcoinPrice(Math.round(1 / response.data.rates.BTC));
        setBitcoinCashPrice(Math.round(1 / response.data.rates.BCH));
        setGoldPrice(Math.round(1 / response.data.rates.XAU));
        setSilverPrice(Math.round(1 / response.data.rates.XAG));
      })
      .catch((error) => {
        if (error.response) {
          // Error if error in retrieving data.
          console.log(error.response.data);
        } else if (error.request) {
          // Error if no response was received.
          console.log(error.request);
        } else {
          // Other errors.
          console.log(error.message);
        }
      });
  };

  return (
    <div className="asset-price-box">
      <div className="title">Current Asset Prices</div>
      {dateBuilder(new Date())}

      <table>
        <tr>
          <th>Bitcoin:</th>
          <td>${bitcoinPrice}</td>
        </tr>
        <tr>
          <th>Bitcoin Cash:</th>
          <td>${bitcoinCashPrice}</td>
        </tr>
        <tr>
          <th>Gold:</th>
          <td>${goldPrice}</td>
        </tr>
        <tr>
          <th>Silver:</th>
          <td>${silverPrice}</td>
        </tr>
      </table>

      <button
        className="price-btn"
        data-testid="get-price-test"
        onClick={fetchAPI}
      >
        <p className="price-btn-title">Get Price</p>
      </button>
    </div>
  );
};

export default AssetPriceChecker;
