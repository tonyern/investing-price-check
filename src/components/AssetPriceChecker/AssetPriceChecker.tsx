import { useState } from "react";
import axios, { AxiosResponse } from "axios";
import "./asset-price-checker.css";

const AssetPriceChecker = (): JSX.Element => {
  const { REACT_APP_METALS_API_ACCESS_KEY } = process.env;
  const metalsAPI = {
    base: "https://www.metals-api.com/api/latest",
    currency: "USD",
  };

  const [bitcoinPrice, setBitcoinPrice] = useState(0);
  const [bitcoinCashPrice, setBitcoinCashPrice] = useState(0);
  const [goldPrice, setGoldPrice] = useState(0);
  const [silverPrice, setSilverPrice] = useState(0);

  const dateBuilder = (d: Date): string => {
    let months: string[] = [
      "January",
      "February",
      "March",
      "April",
      "May",
      "June",
      "July",
      "August",
      "September",
      "October",
      "November",
      "December",
    ];
    let days: string[] = [
      "Sunday",
      "Monday",
      "Tuesday",
      "Wednesday",
      "Thursday",
      "Friday",
      "Saturday",
    ];

    return `${days[d.getDay()]}, ${d.getDate()} 
      ${months[d.getMonth()]} ${d.getFullYear()}`
  };

  const fetchAPI = (): void => {
    axios
      .get(
        `${metalsAPI.base}?access_key=${REACT_APP_METALS_API_ACCESS_KEY}&base=${metalsAPI.currency}`
      )
      .then((response: AxiosResponse<any>) => {
        console.log(response.data);
        setBitcoinPrice(Math.round(1 / response.data.rates.BTC));
        setBitcoinCashPrice(Math.round(1 / response.data.rates.BCH));
        setGoldPrice(Math.round(1 / response.data.rates.XAU));
        setSilverPrice(Math.round(1 / response.data.rates.XAG));
      })
      .catch((error: any) => {
        if (error.response) {
          // Error if error in retrieving data.
          console.log(error.response.data);
        } else if (error.request) {
          // Error if no response was received.
          console.log(error.request);
        } else {
          // Other errors.
          console.log(error.message);
        }
      });
  };

  return (
    <div className="asset-price-box">
      <div className="title">Current Asset Prices</div>
      {dateBuilder(new Date())}

      <table>
        <tbody>
          <tr>
            <th>Bitcoin:</th>
            <td>${bitcoinPrice}</td>
          </tr>
          <tr>
            <th>Bitcoin Cash:</th>
            <td>${bitcoinCashPrice}</td>
          </tr>
          <tr>
            <th>Gold:</th>
            <td>${goldPrice}</td>
          </tr>
          <tr>
            <th>Silver:</th>
            <td>${silverPrice}</td>
          </tr>
        </tbody>
      </table>

      <button
        className="price-btn"
        data-testid="get-price-test"
        onClick={fetchAPI}
      >
        <p className="price-btn-title">Get Price</p>
      </button>
    </div>
  );
};

export default AssetPriceChecker;
