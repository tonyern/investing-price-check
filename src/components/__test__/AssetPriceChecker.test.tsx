import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import AssetPriceChecker from "../AssetPriceChecker/AssetPriceChecker";

it("Renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<AssetPriceChecker />, div);
});

it("Asset price checker snapshot test", () => {
  const tree = renderer.create(<AssetPriceChecker />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Testing get price button", () => {
  const { getByTestId } = render(<AssetPriceChecker />);
  expect(getByTestId("get-price-test")).toHaveTextContent("Get Price");
});
