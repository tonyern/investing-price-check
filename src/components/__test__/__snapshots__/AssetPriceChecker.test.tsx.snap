// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asset price checker snapshot test 1`] = `
<div
  className="asset-price-box"
>
  <div
    className="title"
  >
    Current Asset Prices
  </div>
  Wednesday, 12 
      May 2021
  <table>
    <tbody>
      <tr>
        <th>
          Bitcoin:
        </th>
        <td>
          $
          0
        </td>
      </tr>
      <tr>
        <th>
          Bitcoin Cash:
        </th>
        <td>
          $
          0
        </td>
      </tr>
      <tr>
        <th>
          Gold:
        </th>
        <td>
          $
          0
        </td>
      </tr>
      <tr>
        <th>
          Silver:
        </th>
        <td>
          $
          0
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="price-btn"
    data-testid="get-price-test"
    onClick={[Function]}
  >
    <p
      className="price-btn-title"
    >
      Get Price
    </p>
  </button>
</div>
`;
